overlord/assertstate/assertstate.go

```diff
@@ -1129,3 +1129,77 @@ func TemporaryDB(st *state.State) *asserts.Database {
 	db := cachedDB(st)
 	return db.WithStackedBackstore(asserts.NewMemoryBackstore())
 }
+
+// ValidationSetsFromModel takes in a model and creates a
+// snapasserts.ValidationSets from any validation sets that the model includes.
+func ValidationSetsFromModel(st *state.State, model *asserts.Model, store snapstate.StoreService, offline bool) (*snapasserts.ValidationSets, error) {
+	var sets []*asserts.ValidationSet
+	save := func(a asserts.Assertion) error {
+		if vs, ok := a.(*asserts.ValidationSet); ok {
+			sets = append(sets, vs)
+		}
+
+		if err := Add(st, a); err != nil {
+			if _, ok := err.(*asserts.RevisionError); ok {
```

Comment:
Sorry for any confusion, but Samuele asked that these functions be moved to another PR. #13345.

To answer your question, I believe this is to allow for the case that the model contains a validation set that has already been added to the database. I see other instances in the code (https://github.com/snapcore/snapd/blob/master/asserts/pool.go?plain=1#L545) that imply that this error type might be returned in situations where the DB already has a revision that is newer than the revision we are adding. I'm not sure how to handle that situation here.

Reply:
Ah, I see, then I need to review that PR first.

Thanks for the explanation regarding the error. I guess that in that case (we are trying to add an assertion with a revision older than one in the system) it makes sense to just ignore the error.

Reply:
But maybe it should be logged though? Otherwise we might not know what's going on if, say, we are doing offline remodeling and we are providing local assertions but the db ones are alredy newer.

Reply:
We discussed how these were generated via `snapcraft edit`, so they were never manually written in their JSON form.